### ISSUE TITLE
sql: long running txn with create table hangs concurrent introspection queries

### ISSUE THREAD

**Issue by rjnn** [cockroachdb/cockroach#20526]
A long-running write query hangs all other SQL queries on separate terminals while it runs.

Reproduction steps:
Load a TPC-H dataset in order to get a dataset.
Start cockroach with a large `--max-sql-memory`. I used (on my laptop):
`./cockroach start --insecure --background --max-sql-memory=4gb`
Then run the following query:

`create table on_the_fly as (select ps_partkey as fly_partkey, ps_suppkey as fly_suppkey, p_comment as fly_comment, p_brand as fly_brand from tpch.part full outer join tpch.partsupp on p_partkey = ps_partkey);`

This essentially creates a large table as a JOIN. This takes a while to complete, as the `create table` ensures this runs on local SQL, so even though I'm joining on the prefix of the tables' primary keys, it doesn't use merge joins (as local SQL does not support merge joins). The max-sql-memory is required to witness the hang as otherwise this will OOM out rapidly.

Now in a separate terminal, run
```
oot@localhost:26257/> show databases;
+--------------------+
|      Database      |
+--------------------+
| crdb_internal      |
| graphs             |
| information_schema |
| pg_catalog         |
| system             |
| tpch               |
| ycsb               |
+--------------------+
(9 rows)

Time: 42.956320014s
```
It hangs until the first query finishes (in this case, with `pq: kv/txn_coord_sender.go:422: transaction is too large to commit: 4000007 intents`, although I guess an `OOM` is also probable given what scalefactor and max SQL memory you're using.)

Assigned to @petermattis for triage. cc @vivekmenezes

**Comment by petermattis** [comment]
@arjunravinarayan Do you know how long the `select` portion of the query takes (with both distsql enabled and disabled)? Curious if this is a problem with a large insert or with running a large read-only query without distsql.

**Comment by arjunravinarayan** [comment]
Upon further testing in order to answer your questions, I found that this hanging happens immediately when the query starts, but only for introspection queries (such as `SHOW DATABASES`).

This is a schema change bug. The `CREATE TABLE` creates some lock on the schema state, that lock is held while the schema change is running, and until that schema change aborts or succeeds, all introspection queries hang.

Other possibly unrelated points:

1. A `SELECT` query as part of a `CREATE TABLE`, is never going to run under DistSQL, since it requires writes in the same transaction. Nevertheless, I turned `distsql = off` in my testing.
2. `select count(*) from [select ps_partkey as fly_partkey, ps_suppkey as fly_suppkey, p_comment as fly_comment, p_brand as fly_brand from tpch.part full outer join tpch.partsupp on p_partkey = ps_partkey];` (the wrapping `count` to prevent paging all the input) takes 26seconds on my computer.
3. A `select count(*) from ycsb.usertable` completed successfully without hanging.

**Comment by arjunravinarayan** [comment]
Do note that many ORMs and SQL tools perform some introspection queries on startup, so would be blocked by this bug, while the command line tool would be able to proceed.

**Comment by arjunravinarayan** [comment]
Assigned @vivekmenezes for triaging this as it's a schema change bug.

**Comment by vivekmenezes** [comment]
I suspect the schema change is writing two intents (table-name -> id map and table descriptor) to the store which don't commit until all the data has been written. The intents locks up scans on the name -> id map table as well as scans on the table descriptor table.

**Comment by vivekmenezes** [comment]
@arjunravinarayan can you still reproduce this bug if you move the call to 
```createDescriptorWithID()``` to after the table data has been inserted. You will need to still call 

```
if desc, ok := descriptor.(*sqlbase.TableDescriptor); ok {
		p.session.tables.addUncommittedTable(*desc)
	}
```
that inserts the table descriptor in the local cache, so that the insert finds the table descriptor in the cache and can run along with the inserts.

Thanks!

**Comment by jordanlewis** [comment]
@arjunravinarayan wait, what happened to your WIP here? This seems like we can push this to 2.1, but I'm curious.

**Comment by vivekmenezes** [comment]
This issue is likely fixable by moving schema change writes to just before the commit to the transaction. I feel it's anyway better to not leave intents uncommitted on the descriptor space for a long time.

One of the prerequisites for doing this is to not require such transactions to be anchored around the table descriptor range just for the purposes of gossiping the system range after commit.

**Comment by vivekmenezes** [comment]
This is going to be quite involved to fix. The main limitation here is we do not support long running schema changes. Therefore, we will need to create the table in the user transaction but the underlying data will have to be backfilled outside of the users transaction albeit at the same timestamp as the CREATE TABLE. I think we can add the table in the ADD state so it is disabled, then add the data, and then make the table public.

**Comment by vivekmenezes** [comment]
Fixing this is the first phase of improvements we are making to CREATE TABLE AS. The fix will involve moving CREATE TABLE AS to run as two transactions. The user transaction will be used to create the descriptor and add it in the ADD state. The second transaction will run asynchronously in a schema changer and will populate all the data for the table.

later the second transaction will be replaced by bulk EXPORT-> IMPORT of data

@dt FYI

**Comment by dt** [comment]
This should be resolved by the fixes @adityamaru27 built for #25828.
